### PR TITLE
ScoreP Fixes

### PR DESCRIPTION
**Pull request by pspoerri**
With this I'm able to compile scorep with CUDA 7.0 and ScoreP 3.0. However I'm not able to run it:
```
Invalid buffer pointer, error stack:
MPI_Irecv(148): MPI_Irecv(buf=(nil), count=24120, MPI_DOUBLE, src=1, tag=5, MPI_COMM_WORLD, request=0x1a3edb0) failed
MPI_Irecv(118): Null buffer pointer

[keschcn-0006:mpi_rank_1][error_sighandler] Caught error: Segmentation fault (signal 11)
srun: error: keschcn-0006: task 0: Exited with exit code 1
srun: Terminating job step 1792402.0
srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
slurmstepd: error: *** STEP 1792402.0 ON keschcn-0006 CANCELLED AT 2017-06-02T12:40:05 ***
srun: error: keschcn-0006: task 1: Segmentation fault (core dumped)
```

### CODE REVIEW

**Comment by jgphpc** [comment]
To be able to use scorep, i need to move up `-lmpi -lmpicxx` before `-lcuda -lcudart ...` otherwise i segfault at runtime. 

```
[keschcn-0002:mpi_rank_0][error_sighandler] Caught error: Segmentation fault (signal 11)
```

### make + scorep :ok: 

```
module purge
module load craype-haswell craype-network-infiniband
module load Score-P/3.0-gmvapich2-17.02_cuda_7.5_gdr
cd src/
make clean
export SCOREP_WRAPPER=on
make -f makefile-kesch
```
* After relinking manually, code will work:

```
export SCOREP_TOTAL_MEMORY=1G
export SCOREP_ENABLE_TRACING=true
export SCOREP_CUDA_ENABLE=1
srun -p debug --gres=gpu:1 -N1 -n1 --ntasks-per-node=1 -t5 \
./GNU4.9.3_MVAPICH2-GDR2.2_CUDAV7.5.17.keschln-0003 
```

* vampir  scorep-20170606_1022_5822281186057301/traces.otf2

You need to find a way to do the same within cmake.

**Comment by pspoerri** [comment]
I manually changed the linker line and executed it by hand:
```
/scratch/spoerrip/proj_comm_overlap_bench/comm_overlap_bench_scorep/build/src> 
/apps/escha/UES/RH6.7/easybuild/software/GCC/4.9.3-binutils-2.25/bin/g++ -L/lib64 -std=c++11 -g
-lpthread CMakeFiles/comm_overlap_benchmark.dir/main.cpp.o 
CMakeFiles/comm_overlap_benchmark.dir/MPIHelper.cpp.o 
CMakeFiles/comm_overlap_benchmark.dir/Repository.cpp.o 
CMakeFiles/comm_overlap_benchmark.dir/Options.cpp.o 
CMakeFiles/comm_overlap_benchmark.dir/comm_overlap_benchmark_generated_HorizontalDiffusionSA.cu.o 
CMakeFiles/comm_overlap_benchmark.dir/comm_overlap_benchmark_generated_Kernel.cu.o 
CMakeFiles/comm_overlap_benchmark.dir/comm_overlap_benchmark_intermediate_link.o 
-o comm_overlap_benchmark 
-Wl,-rpath,/opt/mvapich2/gdr/2.1/cuda7.0/gnu/lib64: /opt/mvapich2/gdr/2.1/cuda7.0/gnu/lib64/libmpicxx.so 
/opt/mvapich2/gdr/2.1/cuda7.0/gnu/lib64/libmpi.so 
/global/opt/nvidia/cudatoolkit/7.0.28/lib64/libcudart_static.a 
-ldl /usr/lib64/librt.so
```
and I'm still getting a segfault when executing the executable.

**Comment by pspoerri** [comment]
Addendum: With this I get a segfault when I run with two jobs, if I run with 1 jobs I get no segfault...

**Comment by pspoerri** [comment]
Summarized: 
* If I change the linker line to your recommendation, I'm able to run with one job, but not with 2.
* With the current setup I get segfaults with 1 and 2 jobs.

**Comment by jgphpc** [comment]
I can get a trace with:

```
srun -p debug --gres=gpu:2 -N1 -n2 --ntasks-per-node=2 \
-t5 ./GNU4.9.3_MVAPICH2-GDR2.2_CUDAV7.5.17.keschln-0003 
```

**Comment by pspoerri** [comment]
ok, with this commit above I'm able to compile with CMake
```
./build_kesch_scorep.sh
```
and run it:
* `G2G=1 jobs=2 ./run_kesch_scorep.sh` works
* `G2G=2 jobs=2 ./run_kesch_scorep.sh` does not work (segfault)

File: CMakeLists.txt [hunk 1]
```diff
@@ -68,6 +68,14 @@ if(MPI_LINK_FLAGS)
     set(CMAKE_LINK_FLAGS ${MPI_LINK_FLAGS} ${CMAKE_LINK_FLAGS})
 endif()
 
+#### FORCE MPI Libraries to front by appending to CMAKE_EXE_LINKER_FLAGS
+if(MPI_CXX_LIBRARIES)
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${MPI_CXX_LINK_FLAGS}")
+    foreach(ITEM ${MPI_CXX_LIBRARIES})
+        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${ITEM}")
+    endforeach()
+endif()
```
Comment (pspoerri) [inline CMakeLists.txt:77]: Here's the relevant change in CMake... I had to integrate to rearrange the mpi linker flags.

**Pull request merged**